client/src/game/ui/selection/shapecontext.vue

```diff
@@ -19,85 +19,139 @@
+import { ServerLocation } from "@/game/comm/types/general";
 import { EventBus } from "@/game/event-bus";
 import { layerManager } from "@/game/layers/manager";
 import { Shape } from "@/game/shapes/shape";
 import { gameStore } from "@/game/store";
-import { inInitiative, initiativeStore } from "../initiative/store";
+import { cutShapes, deleteShapes, pasteShapes } from "../../shapes/utils";
+import { initiativeStore, inInitiative } from "../initiative/store";
 
 @Component({
     components: {
         ContextMenu,
     },
+    computed: {
+        ...mapState("game", ["locations", "assets", "notes"]),
+    },
 })
 export default class ShapeContext extends Vue {
     visible = false;
     x = 0;
     y = 0;
-    shape: Shape | null = null;
+    shapes: Shape[] | null = null;
     get activeLayer(): string {
         const layer = layerManager.getLayer();
         return layer === undefined ? "" : layer.name;
     }
-    open(event: MouseEvent, shape: Shape) {
+    open(event: MouseEvent, shapes: Shape[]) {
         this.visible = true;
         this.x = event.pageX;
         this.y = event.pageY;
-        this.shape = shape;
+        this.shapes = shapes;
         this.$nextTick(() => (<HTMLElement>this.$children[0].$el).focus());
     }
     close() {
         this.visible = false;
-        this.shape = null;
+        this.shapes = null;
     }
     getLayers() {
         return layerManager.layers.filter(l => l.selectable && (gameStore.IS_DM || l.playerEditable));
     }
     getActiveLayer() {
         return layerManager.getLayer();
     }
+    getCurrentLocation() {
+        return gameStore.locationName;
+    }
     getInitiativeWord() {
-        if (this.shape === null) return "";
-        return inInitiative(this.shape.uuid) ? "Show" : "Add";
+        if (this.shapes === null) return "";
+        if (this.shapes.length === 1) {
+            return inInitiative(this.shapes[0].uuid) ? "Show" : "Add";
+        } else {
+            return this.shapes.every(shape => inInitiative(shape.uuid)) ? "Show" : "Add all to";
+        }
+    }
+    hasSingleShape(): boolean {
+        return this.shapes !== null && this.shapes.length === 1;
     }
     setLayer(newLayer: string) {
-        if (this.shape === null) return;
-        this.shape.moveLayer(newLayer, true);
+        if (this.shapes === null) return;
+        this.shapes.forEach(shape => shape.moveLayer(newLayer, true));
+        this.close();
+    }
+    setLocation(newLocation: string) {
+        if (this.shapes === null) return;
+        const layer = layerManager.getLayer()!;
+        layer.selection = this.shapes;
+        cutShapes();
+        // Request change to other location
+        socket.emit("Location.Change", newLocation);
+        socket.once("Location.Set", (_data: Partial<ServerLocation>) => {
+            // Paste when location changes
+            // TODO: Shapes are pasted to map layer independently of where they come from. Fix this
```

Comment:
You could alter the pasteShapes signature to 
```typescript
export function pasteShapes(offsetPosition: boolean = true, targetLayer?: string): Shape[] {
    const layer = layerManager.getLayer(targetLayer);
```
which will default to the current layer for normal ctrl+v, and use a specific layer in this part. When the target layer however does not exist, no shapes will actually be pasted as the function will early return.

Another approach is to simply call `selectLayer(layer.name)` in `layers/manager.ts` before pasting.  If the layer does not exist, selectLayer will basically act as a no-op, and the shapes will thus be pasted on the current active layer of that location (which should be the last active layer for the current user on that location).

Your current approach with calling setLayer after pasting them, gives some small communication overhead, given that you first paste all the shapes to then all move them again.  but otherwise is fairly similar to the selectLayer approach.

In practice currently, every location should have the same set of layers, so the aforementioned problems are not really relevant.  I do plan however on adding more flexibility in the layer system in the future, which could result in different sets of layers to exist across locations, hence the thought exercise. At that point either a dialog needs to appear to choose a target layer, or if possible a target suited for tokens is automatically selected.  But hey that's far away.

Reply:
Calling selectLayer before pasting changes the selected layer in the destination location. This should not be a problem, though.

I will change the pasteShapes signature as suggested, i like it.

Reply:
Isn't this what you would expect though ? Or would you assume to keep the original selected layer selected after the paste?

Location A: active layer 'tokens'; Location B: active layer: 'fow'
Move selection of tokens to location B, they will be properly pasted on location B's tokens layer.

What is the expected active layer after the paste happens in location B? tokens or fow?  I was under the impression it would become tokens as chances are high that you'll want to move the tokens you just pasted around.

Reply:
You are right. Probably even if the paste method gets refactored or if the server does the pasting before the layer is even changed, we might want to change the active layer anyway.